src/awscdk/codecatalyst.ts

```diff
@@ -0,0 +1,342 @@
+- upgrade.yaml (upgrade dependencies)  - comes from projen
+
+* synth -> create artifacts
+* upload cdk assets ->  save assets in s3 (lambda), build container images (push to ECR) -- everything in AWS
+* deploy for each stage that is non-production
+* deploy to prod (manual approval)
+
+TODO:
+- account target -- NOT POSSIBLE as we cannot create environments/accounts/targets
+- manual approval for stages -- DONE
+- IAM role per stage, synth, asset - NOT POSSIBLE as we cannot create environments
+- independend stages (all parallel to each other) after synth&assets -- DONE
+- environments support - DONE
+- steps per stage - preInstall, preSynth, ... - DONE
+
+example: https://github.com/aws-community-dach/event-system-backend
+
+test docgen: https://github.com/open-constructs/aws-cdk-library
+
+
+*/
+
+export interface CodeCatalystIamRoleConfig {
+  /** Default IAM role ARN used if no specific role is provided. */
+  readonly default?: string;
+  /** IAM role ARN for the synthesis step. */
+  readonly synth?: string;
+  /** IAM role ARN for the asset publishing step. */
+  readonly assetPublishing?: string;
+  /** IAM role ARN for the asset publishing step for a specific stage. */
+  readonly assetPublishingPerStage?: { [stage: string]: string };
+  /** IAM role ARNs for different deployment stages. */
+  readonly deployment?: { [stage: string]: string };
+}
+
+export interface CodeCatalystCDKPipelineOptions extends CDKPipelineOptions {
+  readonly iamRoleArns: CodeCatalystIamRoleConfig;
+}
+
+export class CodeCatalystCDKPipeline extends CDKPipeline {
+
+  public readonly needsVersionedArtifacts: boolean;
+
+  private deploymentWorkflowBuilder: WorkflowBuilder;
+  private environments: Map<String, Environment> = new Map();
+  private deploymentStages: string[] = [];
+
+  private readonly bp: Blueprint;
+
+  constructor(app: awscdk.AwsCdkTypeScriptApp, private options: CodeCatalystCDKPipelineOptions) {
+    super(app, options);
+    // see https://github.com/aws/codecatalyst-blueprints/issues/477
+    process.env.CONTEXT_ENVIRONMENTID = 'prod';
+
+    this.bp = new Blueprint({ outdir: '.codecatalyst/workflows' });
+
+    if (this.options.iamRoleArns) {
+      this.createEnvironments();
+    }
+
+    this.deploymentWorkflowBuilder = new WorkflowBuilder(this.bp);
+
+    this.deploymentWorkflowBuilder.setName('deploy');
+    this.deploymentWorkflowBuilder.addBranchTrigger(['main']);
+
+    this.needsVersionedArtifacts = this.options.stages.find(s => s.manualApproval === true) !== undefined;
+
+    this.createSynth();
+    this.createAssetUpload();
+
+    for (const stage of options.stages) {
+      this.createDeployment(stage);
+    }
+
+    for (const stage of (options.independentStages ?? [])) {
+      this.createIndependentDeployment(stage);
+    }
+
+    const yml = new YamlFile(this, '.codecatalyst/workflows/deploy.yaml', {
+      obj: this.deploymentWorkflowBuilder.getDefinition(),
+
+    });
+    yml.synthesize();
```

Comment:
That should not be needed. projen calls all synth methods

Reply:
It doesnt for a YamlFile object - or at least I wasnt able to get it do that... As this is the "base" class implementation... maybe I would need to add this as a member of the class to make projen call it?